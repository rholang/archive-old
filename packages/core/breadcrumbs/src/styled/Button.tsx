import React from 'react';
import Button, { ButtonProps } from '@atlaskit/button';

interface Props extends ButtonProps {
  truncationWidth?: number;
}

export default React.forwardRef<HTMLButtonElement, Props>(
  ({ truncationWidth, ...props }, ref) => (
    <Button
      {...props}
      ref={ref}
      theme={(currentTheme, themeProps) => {
        const { buttonStyles, ...rest } = currentTheme(themeProps);
        return {
          buttonStyles: {
            ...buttonStyles,
            ...(truncationWidth
              ? { maxWidth: `${truncationWidth}px !important` }
              : { flexShrink: 1, minWidth: 0 }),
          },
          ...rest,
        };
      }}
    />
  ),
);
