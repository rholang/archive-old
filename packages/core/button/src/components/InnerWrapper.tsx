/** @jsx jsx */
import { jsx } from '@emotion/core';
import React from 'react';

interface Props {
  children: React.ReactNode;
  fit: boolean;
  onClick?: React.MouseEventHandler;
  testId?: string;
}

export default ({ fit, children, ...rest }: Props) => (
  <span
    css={{
      alignSelf: 'center',
      display: 'inline-flex',
      flexWrap: 'nowrap',
      maxWidth: '100%',
      position: 'relative',
      ...(fit && { width: '100%' }),
      ...(fit && { justifyContent: 'center' }),
    }}
    {...rest}
  >
    {children}
  </span>
);
