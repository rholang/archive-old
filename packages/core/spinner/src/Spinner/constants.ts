export const SIZES_MAP: { [key: string]: number } = {
  xsmall: 8,
  small: 16,
  medium: 24,
  large: 48,
  xlarge: 96,
};

export const DEFAULT_SIZE = SIZES_MAP.small;
