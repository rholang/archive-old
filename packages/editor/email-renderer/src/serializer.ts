import { Fragment } from 'prosemirror-model';

export interface Serializer<T> {
  serializeFragment(
    fragment: Fragment,
    props?: any,
    target?: any,
    key?: string,
  ): T | null;
}

export type MediaImageBase64 = {
  contentId: string;
  contentType: string;
  data: string;
};

export interface SerializeFragmentWithAttachmentsResult {
  result: string | null;
  embeddedImages: MediaImageBase64[];
}

export interface SerializerWithImages<T> extends Serializer<T> {
  serializeFragmentWithImages(
    fragment: Fragment,
    props?: any,
    target?: any,
    key?: string,
  ): SerializeFragmentWithAttachmentsResult | null;
}
