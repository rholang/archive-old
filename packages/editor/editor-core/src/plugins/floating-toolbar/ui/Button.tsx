import * as React from 'react';
import Tooltip from '@atlaskit/tooltip';
import Button from '@atlaskit/button';
import { getButtonStyles, baseStyles } from './styles';

export type ButtonAppearance = 'subtle' | 'danger';
export interface Props {
  title?: string;
  icon?: React.ReactElement<any>;
  iconAfter?: React.ReactElement<any>;
  onClick?: React.MouseEventHandler;
  onMouseEnter?: <T>(event: React.MouseEvent<T>) => void;
  onMouseLeave?: <T>(event: React.MouseEvent<T>) => void;
  selected?: boolean;
  disabled?: boolean;
  appearance?: ButtonAppearance;
  href?: string;
  target?: string;
  children?: React.ReactNode;
  className?: string;
}

export default ({
  title,
  icon,
  iconAfter,
  onClick,
  onMouseEnter,
  onMouseLeave,
  selected,
  disabled,
  href,
  target,
  appearance = 'subtle',
  children,
  className,
}: Props) => {
  return (
    <Tooltip content={title} hideTooltipOnClick={true} position="top">
      <div onMouseEnter={onMouseEnter} onMouseLeave={onMouseLeave}>
        <Button
          className={className}
          theme={(adgTheme, themeProps) => {
            const { buttonStyles, ...rest } = adgTheme(themeProps);
            return {
              buttonStyles: {
                ...buttonStyles,
                ...baseStyles,
                ...(appearance === 'danger' &&
                  getButtonStyles({
                    appearance,
                    state: themeProps.state,
                    mode: themeProps.mode,
                  })),
              },
              ...rest,
            };
          }}
          aria-label={title}
          spacing="compact"
          href={href}
          target={target}
          appearance={appearance}
          aria-haspopup={true}
          iconBefore={icon || undefined}
          iconAfter={iconAfter}
          onClick={onClick}
          isSelected={selected}
          isDisabled={disabled}
        >
          {children}
        </Button>
      </div>
    </Tooltip>
  );
};
