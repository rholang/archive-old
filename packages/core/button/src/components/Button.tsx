/** @jsx jsx */
import { jsx } from '@emotion/core';
import React from 'react';
import memoize from 'memoize-one';
import {
  withAnalyticsEvents,
  withAnalyticsContext,
  createAndFireEvent,
} from '@atlaskit/analytics-next';
import {
  name as packageName,
  version as packageVersion,
} from '../version.json';
import GlobalTheme from '@atlaskit/theme/components';
import { Theme } from '../theme';
import { mapAttributesToState, filterProps, composeRefs } from './utils';
import Content from './Content';
import InnerWrapper from './InnerWrapper';
import IconWrapper from './IconWrapper';
import LoadingSpinner from './LoadingSpinner';
import { ButtonProps, ThemeMode, ThemeProps, ThemeTokens } from '../types';

export type ButtonState = {
  isHover: boolean;
  isActive: boolean;
  isFocus: boolean;
};

export class Button extends React.Component<ButtonProps, ButtonState> {
  static defaultProps = {
    appearance: 'default',
    autoFocus: false,
    isDisabled: false,
    isLoading: false,
    isSelected: false,
    shouldFitContainer: false,
    spacing: 'default',
    type: 'button',
  };

  // ref can be a range of things because we render button, a, span or other React components
  button: React.Ref<HTMLButtonElement> = React.createRef<HTMLButtonElement>();

  // Makes sure we don't call ref every render.
  getComposedRefs = memoize(composeRefs);

  state = {
    isActive: false,
    isFocus: false,
    isHover: false,
  };

  componentDidMount() {
    if (this.props.autoFocus && this.button instanceof HTMLButtonElement) {
      this.button.focus();
    }
  }

  isInteractive = () => !this.props.isDisabled && !this.props.isLoading;

  onMouseEnter = (e: React.MouseEvent<HTMLElement>) => {
    this.setState({ isHover: true });
    if (this.props.onMouseEnter) {
      this.props.onMouseEnter(e);
    }
  };

  onMouseLeave = (e: React.MouseEvent<HTMLElement>) => {
    this.setState({ isHover: false, isActive: false });
    if (this.props.onMouseLeave) {
      this.props.onMouseLeave(e);
    }
  };

  onMouseDown = (e: React.MouseEvent<HTMLElement>) => {
    e.preventDefault();
    this.setState({ isActive: true });
    if (this.props.onMouseDown) {
      this.props.onMouseDown(e);
    }
  };

  onMouseUp = (e: React.MouseEvent<HTMLElement>) => {
    this.setState({ isActive: false });
    if (this.props.onMouseUp) {
      this.props.onMouseUp(e);
    }
  };

  onFocus: React.FocusEventHandler<HTMLButtonElement> = event => {
    this.setState({ isFocus: true });
    if (this.props.onFocus) {
      this.props.onFocus(event);
    }
  };

  onBlur: React.FocusEventHandler<HTMLButtonElement> = event => {
    this.setState({ isFocus: false });
    if (this.props.onBlur) {
      this.props.onBlur(event);
    }
  };

  getElement = () => {
    const { href, isDisabled } = this.props;
    if (href) {
      return isDisabled ? 'span' : 'a';
    }
    return 'button';
  };

  // Swallow click events when the button is disabled
  // to prevent inner child clicks bubbling up.
  onInnerClick: React.MouseEventHandler<HTMLElement> = e => {
    if (!this.isInteractive()) {
      e.stopPropagation();
    }
    return true;
  };

  render() {
    const {
      appearance = 'default',
      children,
      className,
      component: CustomComponent,
      consumerRef,
      iconAfter,
      iconBefore,
      isDisabled = false,
      isLoading = false,
      isSelected = false,
      shouldFitContainer = false,
      spacing = 'default',
      theme = (
        current: (props: ThemeProps) => ThemeTokens,
        props: ThemeProps,
      ) => current(props),
      testId,
      ...rest
    } = this.props;

    const attributes = { ...this.state, isSelected, isDisabled };

    const StyledButton: React.ReactType = CustomComponent || this.getElement();

    const iconIsOnlyChild: boolean = !!(
      (iconBefore && !iconAfter && !children) ||
      (iconAfter && !iconBefore && !children)
    );

    const specifiers = (styles: {}) => {
      if (StyledButton === 'a') {
        return {
          'a&': styles,
        };
      } else if (StyledButton === CustomComponent) {
        return {
          '&, a&, &:hover, &:active, &:focus': styles,
        };
      }
      return styles;
    };

    return (
      <Theme.Provider value={theme}>
        <GlobalTheme.Consumer>
          {({ mode }: { mode: ThemeMode }) => (
            <Theme.Consumer
              mode={mode}
              state={mapAttributesToState(attributes)}
              iconIsOnlyChild={iconIsOnlyChild}
              {...this.props}
            >
              {({ buttonStyles, spinnerStyles }) => (
                <StyledButton
                  {...filterProps(rest, StyledButton)}
                  data-testid={testId}
                  ref={this.getComposedRefs(this.button, consumerRef)}
                  onMouseEnter={this.onMouseEnter}
                  onMouseLeave={this.onMouseLeave}
                  onMouseDown={this.onMouseDown}
                  onMouseUp={this.onMouseUp}
                  onFocus={this.onFocus}
                  onBlur={this.onBlur}
                  disabled={isDisabled}
                  className={className}
                  css={specifiers(buttonStyles)}
                >
                  <InnerWrapper
                    onClick={this.onInnerClick}
                    fit={!!shouldFitContainer}
                  >
                    {isLoading && (
                      <LoadingSpinner
                        spacing={spacing}
                        appearance={appearance}
                        isSelected={isSelected}
                        isDisabled={isDisabled}
                        styles={spinnerStyles}
                      />
                    )}
                    {iconBefore && (
                      <IconWrapper
                        isLoading={isLoading}
                        spacing={spacing}
                        isOnlyChild={iconIsOnlyChild}
                        icon={iconBefore}
                      />
                    )}
                    {children && (
                      <Content
                        isLoading={isLoading}
                        followsIcon={!!iconBefore}
                        spacing={spacing}
                      >
                        {children}
                      </Content>
                    )}
                    {iconAfter && (
                      <IconWrapper
                        isLoading={isLoading}
                        spacing={spacing}
                        isOnlyChild={iconIsOnlyChild}
                        icon={iconAfter}
                      />
                    )}
                  </InnerWrapper>
                </StyledButton>
              )}
            </Theme.Consumer>
          )}
        </GlobalTheme.Consumer>
      </Theme.Provider>
    );
  }
}

const createAndFireEventOnAtlaskit = createAndFireEvent('atlaskit');
const ButtonWithRef = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (props, ref) => <Button {...props} consumerRef={ref} />,
);
ButtonWithRef.displayName = 'Button';

export default withAnalyticsContext({
  componentName: 'button',
  packageName,
  packageVersion,
})(
  withAnalyticsEvents({
    onClick: createAndFireEventOnAtlaskit({
      action: 'clicked',
      actionSubject: 'button',
      attributes: {
        componentName: 'button',
        packageName,
        packageVersion,
      },
    }),
  })(ButtonWithRef),
);
