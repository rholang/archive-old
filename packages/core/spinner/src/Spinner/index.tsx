import React, { Component } from 'react';
import { Transition } from 'react-transition-group';
import styled from 'styled-components';

import { SIZES_MAP, DEFAULT_SIZE } from './constants';
import Container from './styledContainer';
import Svg from './styledSvg';
import { SpinnerProps, SpinnerState } from '../types';

const Outer = styled.span`
  display: inline-block;
  vertical-align: middle;
`;
Outer.displayName = 'Outer';

export default class Spinner extends Component<SpinnerProps, SpinnerState> {
  static defaultProps = {
    delay: 100,
    isCompleting: false,
    invertColor: false,
    onComplete: () => {},
    size: 'medium',
  };

  transitionNode: Transition | null = null;

  constructor(props: SpinnerProps) {
    super(props);
    this.state = {
      phase: '',
    };
  }

  enter = () => {
    const { delay } = this.props;
    if (delay) {
      this.setState({ phase: 'DELAY' });
    } else {
      this.setState({ phase: 'ENTER' });
    }
  };

  idle = () => {
    this.setState({ phase: 'IDLE' });
  };

  exit = () => {
    this.setState({ phase: 'LEAVE' });
  };

  endListener = (node: HTMLElement, done: () => void) => {
    const executeCallback = (event: Event): void => {
      // ignore animation events on the glyph

      if ((event.target as SVGElement).tagName === 'svg') {
        return;
      }
      if (this.state.phase === 'DELAY') {
        this.setState({ phase: 'ENTER' });
        this.endListener(node, done);
      } else {
        done();
      }
      return node && node.removeEventListener('animationend', executeCallback);
    };

    // FIX - jest-emotion doesn't recognise the DOM node so it can't add
    // the eventListener in the @atlaskit/button tests.
    // Should be fixed when we move to emotion@10
    if (node && node.addEventListener) {
      return node.addEventListener('animationend', executeCallback);
    }
    return done();
  };

  validateSize = () => {
    const { size } = this.props;
    const spinnerSize = SIZES_MAP[size] || size;
    return typeof spinnerSize === 'number' ? spinnerSize : DEFAULT_SIZE;
  };

  render() {
    const { phase } = this.state;
    const { delay, invertColor, isCompleting, testId } = this.props;
    const size = this.validateSize();

    const strokeWidth = Math.round(size / 10);
    const strokeRadius = size / 2 - strokeWidth / 2;
    return (
      <Outer>
        <Transition
          addEndListener={this.endListener}
          appear
          in={!isCompleting}
          mountOnEnter
          unmountOnExit
          onEnter={this.enter}
          onEntered={this.idle}
          onExit={this.exit}
          onExited={() => this.props.onComplete()}
          timeout={0}
          ref={node => {
            this.transitionNode = node;
          }}
        >
          <Container
            delay={delay / 1000}
            phase={phase}
            size={size}
            data-testid={testId}
          >
            <Svg
              focusable="false"
              height={size}
              invertColor={invertColor}
              phase={phase}
              size={size}
              viewBox={`0 0 ${size} ${size}`}
              width={size}
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle cx={size / 2} cy={size / 2} r={strokeRadius} />
            </Svg>
          </Container>
        </Transition>
      </Outer>
    );
  }
}
