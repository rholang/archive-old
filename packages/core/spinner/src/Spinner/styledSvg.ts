import styled, { css, keyframes } from 'styled-components';
import { themed } from '@atlaskit/theme/components';
import { N500, N0 } from '@atlaskit/theme/colors';
import { SpinnerPhases } from '../types';
import { SIZES_MAP } from './constants';

type StyleParams = {
  invertColor: boolean;
  phase: SpinnerPhases;
  size: number;
};

const getStrokeWidth = (size: number) => Math.round(size / 10);

const getStrokeCircumference = (size: number) => {
  const strokeWidth = getStrokeWidth(size);
  const strokeRadius = size / 2 - strokeWidth / 2;
  return Math.PI * strokeRadius * 2;
};

/* Define keyframes statically to prevent a perfomance issue in styled components v1 where the keyframes function
 * does not cache previous values resulting in each spinner injecting the same keyframe definition
 * in the DOM.
 * This can be reverted to dynamic keyframes when we upgrade to styled components v2
 */
const keyframeNames: { [key: string]: string } = {
  noop: keyframes`
    from { opacity: 0; }
    to { opacity: 0; }
  `,
  rotate: keyframes`
    to { transform: rotate(360deg); }
  `,
  enterOpacity: keyframes`
    from { opacity: 0; }
    to { opacity: 1; }
  `,
  smallEnterStroke: keyframes`
    from { stroke-dashoffset: ${getStrokeCircumference(SIZES_MAP.small)}px; }
    to { stroke-dashoffset: ${getStrokeCircumference(SIZES_MAP.small) *
      0.8}px; }
  `,
  mediumEnterStroke: keyframes`
    from { stroke-dashoffset: ${getStrokeCircumference(SIZES_MAP.medium)}px; }
    to { stroke-dashoffset: ${getStrokeCircumference(SIZES_MAP.medium) *
      0.8}px; }
  `,
  largeEnterStroke: keyframes`
    from { stroke-dashoffset: ${getStrokeCircumference(SIZES_MAP.large)}px; }
    to { stroke-dashoffset: ${getStrokeCircumference(SIZES_MAP.large) *
      0.8}px; }
  `,
  xlargeEnterStroke: keyframes`
    from { stroke-dashoffset: ${getStrokeCircumference(SIZES_MAP.xlarge)}px; }
    to { stroke-dashoffset: ${getStrokeCircumference(SIZES_MAP.xlarge) *
      0.8}px; }
  `,
};

/* If a standard size is used, we can use one of our statically defined keyframes, otherwise
 * we're forced to dynamically create the keyframe and incur a performance cost.
 */
const getEnterStrokeKeyframe = (size: number) => {
  const standardSizeName = Object.keys(SIZES_MAP).find(
    sizeName => size === SIZES_MAP[sizeName],
  );
  if (standardSizeName) {
    return keyframeNames[`${standardSizeName}EnterStroke`];
  }

  const circumference = getStrokeCircumference(size);
  return keyframes`
    from { stroke-dashoffset: ${circumference}px; }
    to { stroke-dashoffset: ${circumference * 0.8}px; }
  `;
};

const spinnerColor = themed({ light: N500, dark: N0 });
const spinnerColorInverted = themed({ light: N0, dark: N0 });

export const getStrokeColor = ({
  invertColor,
  ...props
}: {
  invertColor?: boolean;
}): string | number =>
  invertColor ? spinnerColorInverted(props) : spinnerColor(props);

export const svgStyles = css`
  ${(props: StyleParams) => {
    const circumference = getStrokeCircumference(props.size);

    const animation = (animProps: {
      size: number;
      phase: SpinnerPhases;
      invertColor: boolean;
    }) => {
      const baseAnimation = '0.86s cubic-bezier(0.4, 0.15, 0.6, 0.85) infinite';
      if (animProps.phase === 'ENTER') {
        return css`
          animation: ${baseAnimation} ${keyframeNames.rotate},
            0.8s ease-in-out ${getEnterStrokeKeyframe(animProps.size)},
            0.2s ease-in-out ${keyframeNames.enterOpacity};
        `;
      }
      return css`
        animation: ${baseAnimation} ${keyframeNames.rotate};
      `;
    };

    return css`
      ${animation}
      fill: none;
      stroke: ${getStrokeColor};
      stroke-dasharray: ${circumference}px;
      stroke-dashoffset: ${circumference * 0.8}px;
      stroke-linecap: round;
      stroke-width: ${getStrokeWidth(props.size)}px;
      transform-origin: center;
    `;
  }};
`;

const Svg = styled.svg`
  ${svgStyles};
`;
Svg.displayName = 'SpinnerSvg';
export default Svg;
