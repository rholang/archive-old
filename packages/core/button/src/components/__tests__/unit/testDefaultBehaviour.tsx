import { mount } from 'enzyme';
import React from 'react';
import cases from 'jest-in-case';
import Spinner from '@atlaskit/spinner';
import * as renderer from 'react-test-renderer';
import serializer, { matchers } from 'jest-emotion';
import Button from '../../Button';
import InnerWrapper from '../../InnerWrapper';

expect.extend(matchers);
expect.addSnapshotSerializer(serializer);

beforeEach(() => {
  jest.clearAllMocks();
});

describe('ak-button/default-behaviour', () => {
  it('button should have type="button" by default', () => {
    const wrapper = mount(<Button />);
    expect(wrapper.find('button').props().type).toBe('button');
  });

  it('should render button if there is no href property', () => {
    const wrapper = mount(<Button />);
    expect(wrapper.find('button').length).toBe(1);
    expect(wrapper.find('a').length).toBe(0);
  });

  it('should render link if href property is set', () => {
    const wrapper = mount(<Button href="test" />);
    expect(wrapper.find('a').length).toBe(1);
    expect(wrapper.find('button').length).toBe(0);
  });

  it('should not render link without href prop, even if the target prop is set', () => {
    const wrapper = mount(<Button target="something" />);
    expect(wrapper.find('a').length).toBe(0);
    expect(wrapper.find('button').length).toBe(1);
  });

  it('should render span when the button is disabled and has href property', () => {
    const wrapper = mount(<Button isDisabled href="test" />);
    expect(wrapper.find('span').first().length).toBe(1);
    expect(wrapper.find('button').length).toBe(0);
    expect(wrapper.find('a').length).toBe(0);
  });

  it("should not render span when the button is disabled, but doesn't have href", () => {
    const wrapper = mount(<Button isDisabled />);
    expect(wrapper.find('StyledSpan').length).toBe(0);
    expect(wrapper.find('button').length).toBe(1);
    expect(wrapper.find('a').length).toBe(0);
  });

  it('should render icon if the prop iconBefore is set', () => {
    const Icon = <div id="icon" />;
    const wrapper = mount(<Button href="test" iconBefore={Icon} />);
    expect(wrapper.contains(Icon)).toBe(true);
  });

  it('should render iconBefore before children', () => {
    const Icon = <div id="icon">icon</div>;
    const wrapper = mount(
      <Button href="test" iconBefore={Icon}>
        button
      </Button>,
    );
    expect(wrapper.text()).toBe('iconbutton');
  });

  it('should render icon if the prop iconAfter is set', () => {
    const Icon = <div id="icon" />;
    const wrapper = mount(<Button href="test" iconAfter={Icon} />);
    expect(wrapper.contains(Icon)).toBe(true);
  });

  it('should render iconAfter after children', () => {
    const Icon = <div id="icon">icon</div>;
    const wrapper = mount(
      <Button href="test" iconAfter={Icon}>
        button
      </Button>,
    );
    expect(wrapper.text()).toBe('buttonicon');
  });

  it('should render button with full container width', () => {
    const wrapper = mount(<Button shouldFitContainer />);
    expect(wrapper.find(InnerWrapper).prop('fit')).toBe(true);
  });

  it('should render button without full container width', () => {
    const wrapper = mount(<Button />);
    expect(wrapper.find(InnerWrapper).prop('fit')).toBe(false);
  });

  it('should be able to render both of the icons', () => {
    const Icon1 = <div id="icon">icon1</div>;
    const Icon2 = <div id="icon">icon2</div>;
    const wrapper = mount(
      <Button href="test" iconBefore={Icon1} iconAfter={Icon2}>
        button
      </Button>,
    );
    expect(wrapper.contains(Icon1)).toBe(true);
    expect(wrapper.contains(Icon2)).toBe(true);
    expect(wrapper.text()).toBe('icon1buttonicon2');
  });

  it('should call onClick handler when link is clicked', () => {
    const spy = jest.fn();
    const wrapper = mount(
      <Button href="test" onClick={spy}>
        button
      </Button>,
    );
    wrapper.find('a').simulate('click');
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('should call onClick handler when button is clicked', () => {
    const spy = jest.fn();
    const wrapper = mount(<Button onClick={spy}>button</Button>);
    wrapper.find('button').simulate('click');
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('should not call onClick handler when button is clicked while loading', () => {
    const spy = jest.fn();
    const wrapper = mount(
      <Button isLoading onClick={spy}>
        button
      </Button>,
    );
    wrapper.find(InnerWrapper).simulate('click');
    expect(spy).toHaveBeenCalledTimes(0);
  });

  it('should render tabIndex attribute when the tabIndex property is set', () => {
    let wrapper = mount(<Button tabIndex={0}>button</Button>);
    expect(wrapper.find('button').is('[tabIndex=0]')).toBe(true);
    wrapper = mount(
      <Button href="#" tabIndex={0}>
        link
      </Button>,
    );
    expect(wrapper.find('a').is('[tabIndex=0]')).toBe(true);
    wrapper = mount(
      <Button tabIndex={0} isDisabled>
        span
      </Button>,
    );
    expect(wrapper.find('button').is('[tabIndex=0]')).toBe(true);
  });

  it('should trigger onFocus handler on focus', () => {
    const spy = jest.fn();
    const wrapper = mount(
      <Button tabIndex={0} onFocus={spy}>
        button
      </Button>,
    );
    const button = wrapper.find('button');
    button.prop('onFocus')!({} as any);
    expect(spy).toHaveBeenCalled();
  });

  it('should respect autofocus', () => {
    const wrapper = mount(
      <Button id="testId123" tabIndex={0} autoFocus>
        button
      </Button>,
    );
    const id = document.activeElement ? document.activeElement.id : null;
    expect(wrapper.find('button').prop('id')).toEqual(id);
  });

  describe('isLoading', () => {
    it('should render the loading spinner when isLoading is true', () => {
      const wrapper = mount(<Button isLoading>Some text</Button>);
      expect(wrapper.find(Spinner).length).toEqual(1);
    });
    it('should not render the loading spinner when isLoading is false', () => {
      const wrapper = mount(<Button>Some text</Button>);
      expect(wrapper.find(Spinner).length).toEqual(0);
    });

    it('set the opacity of the text to 0 when isLoading is true', () => {
      const wrapper = renderer
        .create(<Button isLoading>Some text</Button>)
        .toJSON();
      expect(wrapper).toMatchSnapshot();
    });
  });

  it('should trigger onBlur handler on blur', () => {
    const spy = jest.fn();
    const wrapper = mount(
      <Button tabIndex={0} onBlur={spy}>
        button
      </Button>,
    );
    const button = wrapper.find('button');
    button.prop('onBlur')!({} as any);
    expect(spy).toHaveBeenCalled();
  });

  cases(
    'onMouse* prop is called',
    ({ key }: { key: string }) => {
      const spy = jest.fn();
      const onMouseHandler = { [key]: spy };
      const wrapper = mount(<Button {...onMouseHandler}>Button</Button>);
      const button = wrapper.find('button');
      const event = {
        preventDefault: () => {},
      } as React.MouseEvent<HTMLElement>;
      // @ts-ignore
      button.prop(key)!(event);
      expect(spy).toHaveBeenCalled();
    },
    [
      { key: 'onMouseDown' },
      { key: 'onMouseUp' },
      { key: 'onMouseEnter' },
      { key: 'onMouseLeave' },
    ],
  );
});
