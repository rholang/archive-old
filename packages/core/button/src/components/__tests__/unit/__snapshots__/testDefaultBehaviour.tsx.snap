// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ak-button/default-behaviour isLoading set the opacity of the text to 0 when isLoading is true 1`] = `
.emotion-2 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  max-width: 100%;
  position: relative;
}

.emotion-3 {
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  border-width: 0;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: inherit;
  font-style: normal;
  font-weight: normal;
  max-width: 100%;
  outline: none !important;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
  background: rgba(9,30,66,0.04);
  border-radius: 3px;
  box-shadow: 0 0 0 2px inherit;
  color: #42526E !important;
  cursor: default;
  height: 2.2857142857142856em;
  line-height: 2.2857142857142856em;
  padding: 0 8px;
  -webkit-transition: background 0.1s ease-out,box-shadow 0.15s cubic-bezier(0.47,0.03,0.49,1.38);
  transition: background 0.1s ease-out,box-shadow 0.15s cubic-bezier(0.47,0.03,0.49,1.38);
  -webkit-transition-duration: 0.1s,0.15s;
  transition-duration: 0.1s,0.15s;
  vertical-align: middle;
  width: auto;
  pointer-events: none;
}

.emotion-3::-moz-focus-inner {
  border: 0;
  margin: 0;
  padding: 0;
}

.emotion-3:hover {
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: absolute;
  left: 50%;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  margin: 0 4px;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
  opacity: 0;
}

.emotion-2 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  max-width: 100%;
  position: relative;
}

.emotion-3 {
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  border-width: 0;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: inherit;
  font-style: normal;
  font-weight: normal;
  max-width: 100%;
  outline: none !important;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
  background: rgba(9,30,66,0.04);
  border-radius: 3px;
  box-shadow: 0 0 0 2px inherit;
  color: #42526E !important;
  cursor: default;
  height: 2.2857142857142856em;
  line-height: 2.2857142857142856em;
  padding: 0 8px;
  -webkit-transition: background 0.1s ease-out,box-shadow 0.15s cubic-bezier(0.47,0.03,0.49,1.38);
  transition: background 0.1s ease-out,box-shadow 0.15s cubic-bezier(0.47,0.03,0.49,1.38);
  -webkit-transition-duration: 0.1s,0.15s;
  transition-duration: 0.1s,0.15s;
  vertical-align: middle;
  width: auto;
  pointer-events: none;
}

.emotion-3::-moz-focus-inner {
  border: 0;
  margin: 0;
  padding: 0;
}

.emotion-3:hover {
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: absolute;
  left: 50%;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  margin: 0 4px;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
  opacity: 0;
}

.c1 {
  -webkit-animation: 1s ease-in-out forwards cZJnvt;
  animation: 1s ease-in-out forwards cZJnvt;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 24px;
  width: 24px;
}

div + .c1 {
  display: none;
}

.c2 {
  -webkit-animation: 0.86s cubic-bezier(0.4,0.15,0.6,0.85) infinite ibwhYF;
  animation: 0.86s cubic-bezier(0.4,0.15,0.6,0.85) infinite ibwhYF;
  fill: none;
  stroke: #42526E;
  stroke-dasharray: 69.11503837897544px;
  stroke-dashoffset: 55.29203070318036px;
  stroke-linecap: round;
  stroke-width: 2px;
  -webkit-transform-origin: center;
  -ms-transform-origin: center;
  transform-origin: center;
}

.c0 {
  display: inline-block;
  vertical-align: middle;
}

<button
  autoFocus={false}
  className="emotion-3"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  type="button"
>
  <span
    className="emotion-2"
    onClick={[Function]}
  >
    <div
      className="emotion-0"
    >
      <span
        className="c0"
      >
        <span
          className="c1"
          size={24}
        >
          <svg
            className="c2"
            focusable="false"
            height={24}
            size={24}
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx={12}
              cy={12}
              r={11}
            />
          </svg>
        </span>
      </span>
    </div>
    <span
      className="emotion-1"
    >
      Some text
    </span>
  </span>
</button>
`;
